src/libAtomVM/memory.c

```diff
@@ -445,3 +465,83 @@ HOT_FUNC static term memory_shallow_copy_term(term t, term **new_heap, int move)
         abort();
     }
 }
+
+
+void *memory_create_refc(size_t size, const void *data)
+{
+    size_t n = sizeof(struct RefcBinary) + size;
+    struct RefcBinary *refc = malloc(n);
+    if (IS_NULL_PTR(refc)) {
+        fprintf(stderr, "memory_create_refc_binary: Unable to allocate %zu bytes for refc_binary.\n", n);
+        abort();
+    }
+    list_init((struct ListHead *) refc);
+    refc->ref_count = 1;
+    refc->size = size;
+    refc->data = ((char *) refc) + sizeof(struct RefcBinary);
+    if (data != NULL) {
+        memcpy((void *) refc->data, data, size);
+    }
+
+    return refc;
+}
+
+const void *memory_get_refc_data(void *ptr)
```

Comment:
Yeah, good point.  I also think we should implement something close to this via an `erlang:memory` Nif.  I think ERTS will only report a number of bytes used by binaries.  This is a bit more fine grained, but honestly, I don't think the user has much interest in the details.  Mostly users (and our tests) just need to know how much data is being used.  What do you think?

Reply:
+1

Reply:
I also think we should implement something close to this via an erlang:memory Nif instead of via the system_info Nif). I think ERTS will only report a number of bytes used by binaries. This is a bit more fine grained, but honestly, I don't think the user has much interest in the details. Mostly users (and our tests) just need to know how much data is being used. What do you think?

Reply:
See `refc_binary.c`